packages/controllers/src/permissions/Caveat.ts

```diff
@@ -0,0 +1,257 @@
+ */
+export type CaveatDecorator<ParentCaveat extends CaveatConstraint> = (
+  decorated: AsyncRestrictedMethod<RestrictedMethodParameters, Json>,
+  caveat: ParentCaveat,
+) => AsyncRestrictedMethod<RestrictedMethodParameters, Json>;
+
+/**
+ * TODO:docs
+ */
+type ExtractCaveatValueFromDecorator<Decorator extends CaveatDecorator<any>> =
+  Decorator extends (
+    decorated: any,
+    caveat: infer ParentCaveat,
+  ) => AsyncRestrictedMethod<any, any>
+    ? ParentCaveat extends CaveatConstraint
+      ? ParentCaveat['value']
+      : never
+    : never;
+
+/**
+ * TODO:docs
+ */
+export type CaveatValidator<ParentCaveat extends CaveatConstraint> = (
+  caveat: { type: ParentCaveat['type']; value: unknown },
+  origin?: string,
+  target?: string,
+) => void;
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationConstraint = {
+  /**
+   * The string type of the caveat.
+   */
+  type: string;
+
+  /**
+   * The decorator function used to apply the caveat to restricted method
+   * requests.
+   */
+  decorator: CaveatDecorator<any>;
+
+  /**
+   * The validator function used to validate caveats of the associated type
+   * whenever they are instantiated. Caveat are instantiated whenever they are
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<any>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecification<SpecifiedCaveat extends CaveatConstraint> = {
+  /**
+   * The string type of the caveat.
+   */
+  type: SpecifiedCaveat['type'];
+
+  /**
+   * The decorator function used to apply the caveat to restricted method
+   * requests.
+   */
+  decorator: CaveatDecorator<SpecifiedCaveat>;
+
+  /**
+   * The validator function used to validate caveats of the associated type
+   * whenever they are instantiated. Caveat are instantiated whenever they are
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<SpecifiedCaveat>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationMap<
+  Specification extends CaveatSpecificationConstraint,
+> = {
+  [Key in Specification['type']]: Specification extends CaveatSpecification<
+    Caveat<Key, any>
+  >
+    ? Specification
+    : never;
+};
+
+/**
+ * TODO:docs
+ */
+export type ExtractCaveats<
+  Specification extends CaveatSpecification<CaveatConstraint>,
+> = Specification extends any
```

Comment:
At a glance, I think it's a distributive types thing. Should we resolve this or do anything?

Reply:
It would be great to resolve it, but, I was unable to. Maybe we can leave a TODO comment? I'm sure we could get to the bottom of it if we took the time to get a minimal repro, but that'd take longer than I want to spend on this right now.

Reply:
3014a63